File: src/lib/gmp_lib/src/TrajScale/TrajScale.cpp
#include <gmp_lib/TrajScale/TrajScale.h>

#include <gmp_lib/TrajScale/TrajScale_Prop.h>
#include <gmp_lib/TrajScale/TrajScale_Rot_min.h>
#include <gmp_lib/TrajScale/TrajScale_Rot_wb.h>

namespace as64_
{

namespace gmp_
{


TrajScale::TrajScale(unsigned n_dof)
{
  this->Y0d = arma::vec().zeros(n_dof);
  this->Ygd = arma::vec().ones(n_dof);
  //Initialization required for 'setScaleMethod'
  //will be overwirtten then on the 'setInitFinalPos' call
  this->Y0 = arma::vec().zeros(n_dof);
  this->Yg = arma::vec().ones(n_dof);
}


void TrajScale::setNewStartFinalPos(const arma::vec &Y0, const arma::vec &Yg)
{
  this->Y0 = Y0;
  this->Yg = Yg;

  this->T_sc = this->calcScaling();
  this->inv_T_sc = this->calcInvScaling();
}

void TrajScale::setNominalStartFinalPos(const arma::vec &Y0d, const arma::vec &Ygd)
{
  this->Y0d = Y0d;
  this->Ygd = Ygd;

  this->T_sc = this->calcScaling();
  this->inv_T_sc = this->calcInvScaling();
}

gmp_::TrajScale::Ptr TrajScale::deepCopy() const
{
  gmp_::TrajScale::Ptr cp_obj;

  gmp_::TrajScale::ScaleType sc_t = this->getScaleType();
  switch (getScaleType())
  {
    case gmp_::TrajScale::PROP_SCALE:
      cp_obj.reset( new gmp_::TrajScale_Prop(numOfDoFs()) );
      break;
    case gmp_::TrajScale::ROT_MIN_SCALE:
      cp_obj.reset( new gmp_::TrajScale_Rot_min() );
      break;
    case gmp_::TrajScale::ROT_WB_SCALE:
      cp_obj.reset( new gmp_::TrajScale_Rot_wb() );
      arma::vec n_wb = dynamic_cast<const gmp_::TrajScale_Rot_wb *>(this)->getWorkBenchNormal();
      dynamic_cast<gmp_::TrajScale_Rot_wb *>(cp_obj.get())->setWorkBenchNormal(n_wb);
      break;
  }

  cp_obj->setNominalStartFinalPos(Y0d, Ygd);
  cp_obj->setNewStartFinalPos(Y0, Yg);

  return cp_obj;
}

} // namespace gmp_

} // namespace as64_

File: src/matlab/lib/gmp_lib/TrajScale/TrajScale.m
%% Trajectory scaling class
%  Calculates the scaling matrix for scaling spatially a trajectory from a  
%  nominal start/final position {Y0d, Ygd} to a new start/final position {Y0, Yg}.
%  The scaling is performed according to the scaling method (see @TrajScale::ScaleMethod enum)
%  The scaling is returned as matrix T_sc that premultiplies a position to
%  scale it, i.e. Y_scaled = T_sc * Y
%  The scaling is also invertible, i.e. Y = inv(T_sc) * Y_scaled

classdef TrajScale < matlab.mixin.Copyable
    
    properties (Constant, Access = public)
        
        % enum TrajScaleType{
        PROP_SCALE = 0
        ROT_MIN_SCALE = 1
        ROT_WB_SCALE = 2
        %}
        
    end
    
    methods (Access = public)
        
        %% Constructor.
        %  @param[in] n_dof: degrees of freedom.
        function this = TrajScale(n_dof)
            
            this.Y0d = zeros(n_dof,1);
            this.Ygd = ones(n_dof,1);
            % Initialization required for 'setScaleMethod'
            % will be overwirtten then on the 'setInitFinalPos' call
            this.Y0 = zeros(n_dof,1); 
            this.Yg = ones(n_dof,1);
            
            this.setNominalStartFinalPos(this.Y0d, this.Ygd);
            this.setNewStartFinalPos(this.Y0, this.Yg);
            
        end
        
        %% Returns the matrix that scales to the new init/target position, according to the set scaling method.
        %  @return: The scaling matrix to the new init/target position.
        function T_sc = getScaling(this)
            
            T_sc = this.T_sc;
            
        end
        
        %% Returns the inverse of the matrix that scales to the new init/target position, according to the set scaling method.
        %  @return: The inverse scaling matrix to the new init/target position.
        function inv_T_sc = getInvScaling(this)
            
            inv_T_sc = this.inv_T_sc;
            
        end

        %% Set the new start and final position.
        %  @param[in] Y0: new start position.
        %  @param[in] Yg: new final position.
        function setNewStartFinalPos(this, Y0, Yg)
            
            this.Y0 = Y0;
            this.Yg = Yg;
            
            this.T_sc = this.calcScaling();
            this.inv_T_sc = this.calcInvScaling();
            
        end
        
        %% Set the nominal start and final position.
        %  @param[in] Y0d: nominal start position.
        %  @param[in] Ygd: nominal final position.
        function setNominalStartFinalPos(this, Y0d, Ygd)
            
            this.Y0d = Y0d;
            this.Ygd = Ygd;
            
            this.T_sc = this.calcScaling();
            this.inv_T_sc = this.calcInvScaling();
            
        end
        
        function n_dofs = numOfDoFs(this)
            
            n_dofs = length(this.Y0d);
            
        end
        
        function cp_obj = deepCopy(this)
           
            cp_obj = this.copy();
            
        end
        
    end
    
    methods (Abstract, Access = public)
        
        scale_type = getScaleType(this)
        
    end
    
    methods (Abstract, Access = protected)
        
        sc = calcScaling(this)
        
        inv_sc = calcInvScaling(this)
        
    end
    
    properties (Access = protected)

        Y0d % Nominal start position.
        Ygd % Nominal final position.
        Y0 % New start position.
        Yg % New final position.
        
        T_sc % scaling matrix, calculated based on the set scale method
        inv_T_sc % inverse scaling matrix, calculated based on the set scale method

    end
    
end
